src/libmain.c

```diff
@@ -765,16 +769,20 @@ static gint setup_config_dir(void)
 	return mkdir_result;
 }
 
-/* Signal handling removed since main_quit() uses functions that are
- * illegal in signal handlers
-static void signal_cb(gint sig)
+
+#ifdef G_OS_UNIX
+static gboolean signal_cb(gpointer user_data)
 {
+	gint sig = GPOINTER_TO_INT(user_data);
 	if (sig == SIGTERM)
 	{
+		geany_debug("Received SIGTERM signal");
 		main_quit();
 	}
+	return G_SOURCE_REMOVE;
```

Comment:
If Geany ever returns to the main loop during the quit (which at least GIO file saves would do I think) then you should remove it to avoid getting it again.  But if you are going to allow multiple signals in the one handler then you may want to continue and then you need code to filter that signal out since you don't want to re-enter `main_quit()` or *bad things* will happen.  There is already a quitting global you could use for that :)